Multi-tracking Authorization Edit

sanjaytkbabu opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
# Description
[PADS-565](https://apps.nrs.gov.bc.ca/int/jira/browse/PADS-565) - Multi-auth edit
<!-- Describe your changes in detail -->

Backend:
Permit route, controller, service updates - createPermit/updatePermit replaced with upsertPermit
Permit validator updated
Api spec updated
Tests updated

Frontend:
Authorization edit functionality added
Authorization form updated to use permit store to get permits and pass them to authorization components
en-CA updated
Unused service permitNoteService.updatePermit removed
Tests updated 

<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->

## Types of changes

<!-- What types of changes does your code introduce? Uncomment all that apply: -->

<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Documentation (non-breaking change with enhancements to documentation) -->
<!-- Breaking change (fix or feature that would cause existing functionality to change) -->

## Checklist

<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have checked that unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

<!-- If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc... -->

slhurley commented:
- If you go to add an ID and try to save without touching the status update, it seems to think you need to update the note but you don't as the status has not changed.
- If you click in the breadcrumbs after making any change it does not tell you have not saved yet and loses the info.  Same if you use the back button. I think we need the "unsaved changes" warning.
- Now that we have turned this into a page, I feel the breadcrumbs need to allow them to go back to the Auth tab?
- I don't think you should be able to select the exact same tracking id more than once.

naixin-zhangbc commented:
- After a new authorization is added, the tracking IDs of the existing authorization get removed somehow. 
- When trying to change the "Needed" field from "Under investigation" to "Yes" or "No", it shows "Error: Failed to save permit. Request failed with status code 422"
- Not part of this ticket... just wanted to mention in case it's related to other issues... the Submitted Date doesn't show in the notification email when an update note is added to an authorization.
- Please update the grey background for the "agency, business domain..." row and the "update note history" to FAF9F8

sanjaytkbabu commented:
> * After a new authorization is added, the tracking IDs of the existing authorizations get removed somehow.

fixed!
> * When trying to change the "Needed" field from "Under investigation" to "Yes" or "No", it shows "Error: Failed to save permit. Request failed with status code 422"

It was caused by missing status verification date, it's been fixed as well.
> * Not part of this ticket... mentioning it in case it's related to other issues... the Submitted Date doesn't show up in the notification email. The email that's sent to Prop when an update note is added to an authorization.

Thank you for mentioning this, it happened due to date being returned in a different format. Fixed!
> * Wondering if it's possible to show the "Authorization type and IDs" section expanded, when the authorization come in through Props' intake form and when it's the first time Navs open it up. Asking because that's when Navs need to use that section to add in Tracking IDs.

Will get back to you on this.
> * Please update the grey background for the "agency, business domain..." row and the "update note history" to FAF9F8

done!

sanjaytkbabu commented:
> * If you go to add an ID and try to save without touching the status update, it seems to think you need to update the note but you don't as the status has not changed.

done!
> * If you click in the breadcrumbs after making any change it does not tell you have not saved yet and loses the info.  Same if you use the back button. I think we need the "unsaved changes" warning.
> * Now that we have turned this into a page, I feel the breadcrumbs need to allow them to go back to the Auth tab?
> * I don't think you should be able to select the exact same tracking id more than once.

slhurley commented:
- In the bottom where Business Domain and Agency is shown, can we have a space between the label and the value?
- One thing that is a bit of a problem is when the prop fills out the tracking id on the intake form it is not linked to any particular tracking id name when it is saved.  I guess this is because we have not linked names to particular permits.  We will need to think about how to handle this in future.

Subin1Doo commented:
Can we also vertically left align "Agency" with the "Update note history" header while keeping the spacing equally distributed between Agency, business domain, updated by, and last updated.

Subin1Doo commented:
the date and time format needs commas according to the BC guidelines, we should stick to this from now on,, so format should be [December 31, 2016] for just the date, date and time [September 25, 2015, 1:30 pm]. Thanks!

Subin1Doo commented:
<img width="1715" height="321" alt="Screenshot 2025-07-18 at 11 53 56 AM" src="https://github.com/user-attachments/assets/83f83c9e-f1c0-4b26-ae1b-e1bbe5be8760" />
We need some spacing between the 'update' text section and permit numbers box.

sanjaytkbabu commented:
> * In the bottom where Business Domain and Agency is shown, can we have a space between the label and the value?

done!
> * One thing that is a bit of a problem is when the prop fills out the tracking id on the intake form it is not linked to any particular tracking id name when it is saved.  I guess this is because we have not linked names to particular permits.  We will need to think about how to handle this in future.

jujaga commented on app/src/controllers/housingProject.ts:
```diff
@@ -240,8 +240,8 @@ const controller = {
       // Create each permit
-      await Promise.all(appliedPermits.map(async (x: Permit) => await permitService.createPermit(x)));
-      await Promise.all(investigatePermits.map(async (x: Permit) => await permitService.createPermit(x)));
+      await Promise.all(appliedPermits.map(async (x: Permit) => await permitService.upsertPermit(x)));
+      await Promise.all(investigatePermits.map(async (x: Permit) => await permitService.upsertPermit(x)));
```
The map callback function does not need an async and await - the permitService should be a Promise function already and should be handled by Promise.all's collection.

jujaga commented on app/src/docs/v1.api-spec.yaml:
```diff
@@ -949,7 +916,7 @@ paths:
       responses:
         '200':
-          description: Permit updated successfully
+          description: Permit upserted successfully
```
Will be something to converse about, but generally API documentation focuses on the outcome of the action, not how it was implemented. In this case, it may make sense to retain the updated instead of changing to upserted, because it is fundamentally an update, only that you achieve it with an upsert operation instead.

jujaga commented on app/src/services/permit.ts:
```diff
@@ -173,22 +134,79 @@ const service = {
+              notIn: permitDataWithId.permitTracking
+                ?.map((x: PermitTracking) => x.permitTrackingId)
+                .filter((x) => x) as number[]
+            }
+          }
+        });
+
+        // Upsert the permit tracking
+        if (permitDataWithId.permitTracking?.length) {
+          await Promise.all(
+            permitDataWithId.permitTracking.map(async (x: PermitTracking) => {
+              if (x.permitTrackingId) {
+                await trx.permit_tracking.update({
+                  where: {
+                    permitTrackingId: x.permitTrackingId
+                  },
+                  data: {
+                    permitId: permitDataWithId.permitId,
+                    trackingId: x.trackingId,
+                    sourceSystemKindId: x?.sourceSystemKindId,
+                    shownToProponent: x?.shownToProponent ?? false,
+                    updatedBy: permitDataWithId.updatedBy
+                  }
+                });
+              } else {
+                await trx.permit_tracking.create({
+                  data: {
+                    permitId: permitDataWithId.permitId,
+                    trackingId: x.trackingId,
+                    sourceSystemKindId: x?.sourceSystemKindId,
+                    shownToProponent: x?.shownToProponent ?? false,
+                    createdBy: permitDataWithId.createdBy,
+                    updatedBy: permitDataWithId.updatedBy
+                  }
+                });
+              }
+            })
+          );
         }
+        return response;
```
Two observations:
- The three main blocks of operations may be await Promise.all wrapped together - I'm not seeing any immediate ordering constraints on the three operations (aka something must execute first as the next call is dependent on the result)
- The result of trx.permit.upsert is only used on the return - may be good idea to do this operation last, as then you could just return that action instead of declaring another const just to use it in one place.
There's an opportunity to streamline the async subfunctions here.

jujaga commented on app/src/services/permit.ts:
```diff
@@ -173,22 +134,79 @@ const service = {
+            permitId: permitDataWithId.permitId,
+            permitTrackingId: {
+              notIn: permitDataWithId.permitTracking
+                ?.map((x: PermitTracking) => x.permitTrackingId)
+                .filter((x) => x) as number[]
```
If possible, we should try to avoid using Typescript as assertions, as they should be considered tools of last resort. You should be able to cast this earlier in the chain as a number like this, which properly asserts the return type of map and cascades further down:

```ts
?.map<number>((x: PermitTracking) => x.permitTrackingId)
.filter((x) => x)
```

jujaga replied:
In that case, you need to do one of the following:
- Declare that x.permitTrackingId is not going to be a null value with `x.permitTrackingId!` with trailing exclamation
- `?.map<number | undefined>((x: PermitTracking) => x.permitTrackingId)`

jujaga commented on frontend/src/components/authorization/AuthorizationForm.vue:
```diff
@@ -1,67 +1,72 @@
+const { getProject, getPermits } = storeToRefs(projectStore);
+
+// State
+const authorization: Ref<Permit | undefined> = ref(undefined);
+const formRef: Ref<InstanceType<typeof Form> | null> = ref(null);
+const initialFormValues: Ref<any | undefined> = ref(undefined);
+const authorizationNotes: Ref<Array<PermitNote> | undefined> = ref([]);
+const updatedBy: Ref<User | undefined> = ref(undefined);
```
Alphabetical ordering

jujaga commented on frontend/src/components/authorization/AuthorizationForm.vue:
```diff
@@ -155,6 +190,44 @@ const formSchema = object({
+        authorizationType: authorization.value.permitType,
+        adjudicationDate: authorization.value.adjudicationDate ? new Date(authorization.value.adjudicationDate) : null,
+        submittedDate: authorization.value.submittedDate ? new Date(authorization.value.submittedDate) : null,
+        /* eslint-disable indent */
```
We should not be overriding eslint like this without a very good reason otherwise. I don't see a companion enable clause either.

kyle1morel replied:
If this is a prettier formatting issue use the `prettier-ignore` directive instead.

sanjaytkbabu replied:
tried prettier-ignore, but got error `Expected indentation of 12 spaces but found 14.eslint[indent](https://eslint.org/docs/latest/rules/indent)`

jujaga commented on frontend/tests/unit/service/permitService.spec.ts:
```diff
@@ -59,7 +59,7 @@ vi.mocked(appAxios).mockReturnValue({
 
 // Spies
-const updatePermitSpy = vi.spyOn(permitService, 'updatePermit');
+const upsertPermitSpy = vi.spyOn(permitService, 'upsertPermit');
```
Please make sure to address any test and lint warnings as you see them.

sanjaytkbabu replied:
Had commented out a test to deal with it later. I just fixed it!

kyle1morel commented on app/src/services/permit.ts:
```diff
@@ -173,22 +134,79 @@ const service = {
+          }
+        });
+
+        // Upsert the permit tracking
+        if (permitDataWithId.permitTracking?.length) {
+          await Promise.all(
+            permitDataWithId.permitTracking.map(async (x: PermitTracking) => {
+              if (x.permitTrackingId) {
+                await trx.permit_tracking.update({
+                  where: {
```
Upsert calls for `permit_tracking` table does not belong in `permit` service.

kyle1morel commented on frontend/src/components/authorization/AuthorizationCard.vue:
```diff
@@ -81,30 +90,50 @@ watchEffect(() => {
                 class="mr-2"
                 icon="fa-solid fa-eye"
               />
-              <span class="font-bold">
-                {{ trackingShownToProponent?.sourceSystemKind?.description }}
+              <span
+                v-tooltip.right="t('authorization.authorizationCard.clickToCopy')"
+                class="cursor-pointer"
+                @click="toCopy(trackingShownToProponent?.trackingId as string)"
+              >
+                <span class="font-bold">
+                  {{ trackingShownToProponent?.sourceSystemKind?.description }}
+                </span>
+                :
+
+                {{ trackingShownToProponent?.trackingId }}
```
`?` should be unnecessary after `trackingShownToProponent` as this whole thing is warpped in a `v-if` on L86.

kyle1morel commented on frontend/src/components/authorization/AuthorizationCard.vue:
```diff
@@ -81,30 +90,50 @@ watchEffect(() => {
-              <span class="font-bold">{{ permitTracking.sourceSystemKind?.description }}</span>
-              :
-              {{ permitTracking?.trackingId }}
+              <span
+                v-tooltip.right="t('authorization.authorizationCard.clickToCopy')"
+                class="cursor-pointer"
+                @click="toCopy(permitTracking?.trackingId as string)"
+              >
+                <span class="font-bold">{{ permitTracking.sourceSystemKind?.description }}</span>
+                :
+                {{ permitTracking?.trackingId }}
+              </span>
```
Unnecessary `?` after `permitTracking`

kyle1morel commented on frontend/src/components/authorization/AuthorizationCard.vue:
```diff
@@ -81,30 +90,50 @@ watchEffect(() => {
+              <span
+                v-tooltip.right="t('authorization.authorizationCard.clickToCopy')"
+                class="cursor-pointer"
+                @click="toCopy(permitTracking?.trackingId as string)"
```
Unnecessary casting. `permitTracking` is guaranteed to exist so the `?` can be removed.

kyle1morel commented on frontend/src/components/authorization/AuthorizationCardLite.vue:
```diff
@@ -15,6 +15,9 @@ const { permit } = defineProps<{
 }>();
 
+// Emits
+const emit = defineEmits(['authorizationCard:more']);
```
Emit name does not match file.

kyle1morel commented on frontend/src/components/authorization/AuthorizationForm.vue:
```diff
@@ -1,67 +1,72 @@
+const authorization: Ref<Permit | undefined> = ref(undefined);
+const formRef: Ref<InstanceType<typeof Form> | null> = ref(null);
+const initialFormValues: Ref<any | undefined> = ref(undefined);
+const authorizationNotes: Ref<Array<PermitNote> | undefined> = ref([]);
```
Arrays should not be possibly undefined. Instead favour setting them to an empty array. This prevents `?` checking before calling array functions.

kyle1morel commented on frontend/src/components/authorization/AuthorizationForm.vue:
```diff
@@ -187,39 +267,63 @@ const formSchema = object({
+          @click="
+            router.push({
+              name: projectRouteName,
+              params: { projectId: projectId },
```
I think you can use `params: { projectId },` here. Shouldn't need to redeclare it as the same name.

sanjaytkbabu replied:
this works, I removed it from 3 places in the same file.

sanjaytkbabu commented on app/src/services/permit.ts:
```diff
@@ -173,22 +134,79 @@ const service = {
-        data: { ...permit.toPrismaModel(data), updated_by: data.updatedBy },
-        where: {
-          permit_id: data.permitId
+      const permitDataWithId = { ...data, permitId: data.permitId || uuidv4() };
```
moved the logic to controller layer.

sanjaytkbabu commented:
> * If you click in the breadcrumbs after making any change it does not tell you have not saved yet and loses the info.  Same if you use the back button. I think we need the "unsaved changes" warning.

done!
> * Now that we have turned this into a page, I feel the breadcrumbs need to allow them to go back to the Auth tab?
> * I don't think you should be able to select the exact same tracking id more than once.

I spoke with Kyle, Because of the way we've setup breadcrumbs and tracking information rows being dynamic it'd too difficult and the effort wold be much more than what it's worth.

naixin-zhangbc commented:
> > On the Add/Edit an authorization page
> > 
> > * When a required field is left empty upon publishing, should it trigger an alert toast, or scroll to the required field? Whichever is used elsewhere on Nav side would be good. I see in the Prop intake form, it takes users to the missing fields.
> 
> done!
> 
> > * When the sections are collapsed (Authorization types and IDs, Status updates, etc.), can we make the entire header row clickable to expand the section? So that users don't have to click on the +/- button precisely.
> 
> In talks with Sharolyn, we'd have to do this change across the board to keep things consistent (electrification). We're essentially changing from Panel to Accordion component. Concluded that the effort is more than what it's worth for this ticket.
> 
> > * Thought I saw this been fixed before that by default Authorization Status should show "None", and Application Stage "Pre-submission". When adding a new authorization, these two fields are blank rn.
> 
> Oops, it was missed when we did the edit, put it back!
> 
> > * This is how the business domain row shows up when opening the Add authorization page. Can we get rid of the empty space and comma?
> 
> Removed it on add auth, only appears during edit.

Thank you!

Here's the new issue we discovered

- When adding a new authorization, upon publishing, it shows this message, clears all the filled info, and shows the green success toast in the background (didn't capture it in the screenshot). This seems to only happen when a tracking id field is added.
<img width="1698" height="870" alt="Screenshot 2025-07-23 at 9 37 11 AM" src="https://github.com/user-attachments/assets/35928087-ca16-451d-bb05-22b0111ede0a" />

sanjaytkbabu commented:
> > > On the Add/Edit an authorization page
> > > 
> > > * When a required field is left empty upon publishing, should it trigger an alert toast, or scroll to the required field? Whichever is used elsewhere on Nav side would be good. I see in the Prop intake form, it takes users to the missing fields.
> > 
> > 
> > done!
> > > * When the sections are collapsed (Authorization types and IDs, Status updates, etc.), can we make the entire header row clickable to expand the section? So that users don't have to click on the +/- button precisely.
> > 
> > 
> > In talks with Sharolyn, we'd have to do this change across the board to keep things consistent (electrification). We're essentially changing from Panel to Accordion component. Concluded that the effort is more than what it's worth for this ticket.
> > > * Thought I saw this been fixed before that by default Authorization Status should show "None", and Application Stage "Pre-submission". When adding a new authorization, these two fields are blank rn.
> > 
> > 
> > Oops, it was missed when we did the edit, put it back!
> > > * This is how the business domain row shows up when opening the Add authorization page. Can we get rid of the empty space and comma?
> > 
> > 
> > Removed it on add auth, only appears during edit.
> 
> Thank you!
> 
> Here's the new issue we discovered
> 
> * When adding a new authorization, upon publishing, it shows this message, clears all the filled info, and shows the green success toast in the background (didn't capture it in the screenshot). This seems to only happen when a tracking id field is added.
> 
> <img alt="Screenshot 2025-07-23 at 9 37 11 AM" width="1698" height="870" src="https://private-user-images.githubusercontent.com/189301067/469922120-35928087-ca16-451d-bb05-22b0111ede0a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTMzMDQ1NzMsIm5iZiI6MTc1MzMwNDI3MywicGF0aCI6Ii8xODkzMDEwNjcvNDY5OTIyMTIwLTM1OTI4MDg3LWNhMTYtNDUxZC1iYjA1LTIyYjAxMTFlZGUwYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcyM1QyMDU3NTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00MGQ4NjExM2NmOWZmMTk4Nzg2NDM1ZDYyYjdhZDJlN2I3NWFmMGM1MzcxMmExYjdlOTYwNmNjNWM3NmRkYjllJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.0DKZ5njj_FsOztU8DXNoA58qWR_cUOmWK5TizbM6zik">

must be fixed now

kyle1morel commented on app/tests/unit/controllers/permit.spec.ts:
```diff
@@ -109,11 +112,12 @@ describe('createPermit', () => {
     expect(createSpy).toHaveBeenCalledTimes(1);
     expect(createSpy).toHaveBeenCalledWith({
       ...req.body,
+      permitId: expect.any(String),
```
`expect.stringMatching(uuidv4Pattern)`

kyle1morel commented on app/tests/unit/controllers/permit.spec.ts:
```diff
@@ -393,11 +397,12 @@ describe('updatePermit', () => {
     expect(updateSpy).toHaveBeenCalledTimes(1);
     expect(updateSpy).toHaveBeenCalledWith({
       ...req.body,
+      permitId: expect.any(String),
```
`expect.stringMatching(uuidv4Pattern)`

kyle1morel commented on app/tests/unit/controllers/permit.spec.ts:
```diff
@@ -428,11 +433,12 @@ describe('updatePermit', () => {
     expect(updateSpy).toHaveBeenCalledTimes(1);
     expect(updateSpy).toHaveBeenCalledWith({
       ...req.body,
+      permitId: expect.any(String),
```
`expect.stringMatching(uuidv4Pattern)`

kyle1morel commented on frontend/src/components/authorization/AuthorizationCard.vue:
```diff
@@ -36,6 +40,11 @@ watchEffect(() => {
   }
 });
+
+function toCopy(toCopy: string) {
+  navigator.clipboard.writeText(toCopy);
+  toast.info(t('authorization.authorizationCard.copiedToClipboard'));
+}
```
Style: Should be before lifecycle functions

kyle1morel commented on frontend/src/components/authorization/AuthorizationCard.vue:
```diff
@@ -81,30 +90,50 @@ watchEffect(() => {
 </template>
+
+<style lang="scss" scoped>
+.cursor-pointer {
+  cursor: pointer;
+}
```
Should exist in tailwind somewhere

kyle1morel commented on frontend/src/components/authorization/AuthorizationForm.vue:
```diff
@@ -1,67 +1,77 @@
 import { storeToRefs } from 'pinia';
 import { Form } from 'vee-validate';
-import { inject } from 'vue';
+import { inject, onBeforeMount, ref } from 'vue';
```
Alphabetize

kyle1morel commented on frontend/src/components/authorization/AuthorizationForm.vue:
```diff
@@ -155,6 +195,66 @@ const formSchema = object({
+}
+
+function onInvalidSubmit(e: any) {
+  const errors = Object.keys(e.errors);
+
+  // Scrolls to the top-most error
+  let first: Element | null = null;
+
+  for (let error of errors) {
+    let el = document.querySelector(`[name="${error}"]`);
+    let rect = el?.getBoundingClientRect();
+
+    if (rect) {
+      if (!first) first = el;
+      else if (rect.top < first.getBoundingClientRect().top) first = el;
+    }
+  }
+
+  first?.scrollIntoView({ behavior: 'smooth' });
```
This bit is used across all our forms. We should turn this into a util function.

kyle1morel commented on frontend/src/utils/utils.ts:
```diff
@@ -274,6 +274,30 @@ export function partition<T>(
+ * Scrolls to the first invalid field in a form
+ * @param {object} e The event object containing validation errors
+ */
+export function scrollToFirstError(e: any) {
```
`any` type should not be used. I believe the errors object is a simple object containing key/values. 

`export function scrollToFirstError(errors: { [key: string]: string })`

We can then also remove line 283, and when this function is used explicitly pass in only the errors instead of the entire event - `scrollToFirstError(e.errors);`
